browser/ai_chat/content_agent_tool_provider.cc

```diff
@@ -0,0 +1,183 @@
+    // certain URLs, e.g. NTP. Safer to start on a blank page
+    // whilst this feature is focused on AI-initiated tasks instead
+    // of acting on existing tabs.
+    NavigateParams params(profile_, GURL(url::kAboutBlankURL),
+                          ui::PAGE_TRANSITION_FROM_API);
+    params.disposition = WindowOpenDisposition::NEW_BACKGROUND_TAB;
+    Navigate(&params);
+    content::WebContents* new_contents = params.navigated_or_inserted_contents;
+
+    task_tab_handle_ =
+        tabs::TabInterface::GetFromContents(new_contents)->GetHandle();
+  }
+  actor_service_->GetTask(task_id_)->AddTab(
+      task_tab_handle_, base::BindOnce(&ContentAgentToolProvider::TabAdded,
+                                       weak_ptr_factory_.GetWeakPtr(),
+                                       task_tab_handle_, std::move(callback)));
+}
+
+void ContentAgentToolProvider::TabAdded(
+    tabs::TabHandle tab_handle,
+    base::OnceCallback<void(tabs::TabHandle)> callback,
+    actor::mojom::ActionResultPtr result) {
+  std::move(callback).Run(tab_handle);
+}
+
+void ContentAgentToolProvider::ExecuteActions(
+    optimization_guide::proto::Actions actions,
+    Tool::UseToolCallback callback) {
+  actor::BuildToolRequestResult requests = actor::BuildToolRequest(actions);
+
+  if (!requests.has_value()) {
+    DLOG(ERROR) << "Action Failed to convert BrowserAction to ToolRequests.";
+    std::move(callback).Run(
+        CreateContentBlocksForText("Action failed - incorrect parameters"));
+    return;
+  }
+
+  actor_service_->PerformActions(
+      actor::TaskId(actions.task_id()), std::move(requests.value()),
+      base::BindOnce(&ContentAgentToolProvider::OnActionsFinished,
+                     weak_ptr_factory_.GetWeakPtr(), std::move(callback)));
+}
+
+void ContentAgentToolProvider::CreateTools() {
+  tools_.clear();
+
+  if (actor_service_) {
+    // Note: TaskId is consistent for this conversation. But we might
+    // want to make this more dynamic and have the the tool use request
+    // specify which task (set of tabs) to act on.
+    tools_.push_back(std::make_unique<NavigationTool>(this, actor_service_));
+  }
+}
+
+void ContentAgentToolProvider::OnActionsFinished(
+    Tool::UseToolCallback callback,
+    actor::mojom::ActionResultCode result_code,
+    std::optional<size_t> index_of_failed_action,
+    std::vector<actor::ActionResultWithLatencyInfo> action_results) {
+  if (result_code == actor::mojom::ActionResultCode::kOk) {
+    // Send current page content for result
+
+    // TODO(https://github.com/brave/brave-browser/issues/49259):
+    // Use multi_source_page_context_fetcher.h (or use it
+    // via ActorKeyedService), now that this API is public outside of glic.
+
+    auto options = blink::mojom::AIPageContentOptions::New();
+    options->mode = blink::mojom::AIPageContentMode::kActionableElements;
+
+    optimization_guide::GetAIPageContent(
+        task_tab_handle_.Get()->GetContents(), std::move(options),
```

Comment:
should we `CHECK` TaskTabHandle?

Reply:
Also, is it true that we want to get the page content after every action? I guess it makes sense